feat(share:git-prompt): be able to truncate the branch name with a de…

Neonox31 opened the pull request:
…fined max length

## Description

Add a new `__fish_git_prompt_branch_max_length` variable to define a max length for shown git branch.

faho commented on share/functions/__fish_git_prompt.fish:
```diff
@@ -417,6 +417,14 @@ function __fish_git_prompt --description "Prompt function for Git"
+     if test -z "$__fish_git_prompt_branch_max_length"
+        set __fish_git_prompt_branch_max_length 20
+    end
+    set -l b_ori_length (string length $b)
```
These probably don't need `--` since a branch name starting with `-` is invalid (`git branch -- -something` says so).

It wouldn't nurt, but it's probably okay to leave it out.

faho commented on share/functions/__fish_git_prompt.fish:
```diff
@@ -417,6 +417,14 @@ function __fish_git_prompt --description "Prompt function for Git"
+        set __fish_git_prompt_branch_max_length 20
+    end
+    set -l b_ori_length (string length $b)
+    set b (string sub -l $__fish_git_prompt_branch_max_length $b)
```
Why not just do this if it's too long (i.e. move this one line down)?

faho commented on share/functions/__fish_git_prompt.fish:
```diff
@@ -417,6 +417,14 @@ function __fish_git_prompt --description "Prompt function for Git"
+    end
+    set -l b_ori_length (string length $b)
+    set b (string sub -l $__fish_git_prompt_branch_max_length $b)
+    if test $b_ori_length -gt $__fish_git_prompt_branch_max_length
```
This should probably both quote the variable and silence stderr - if it has more than one element or isn't an integer, this `test` will fail.

Which is alright (it should probably do what `$fish_prompt_pwd_dir_length` does in that case and not truncate), but it'll also print an error, which isn't.

faho replied:
In general, `test` is the builtin that needs the most care.

faho commented on share/functions/__fish_git_prompt.fish:
```diff
@@ -417,6 +417,14 @@ function __fish_git_prompt --description "Prompt function for Git"
+    set -l b_ori_length (string length $b)
+    set b (string sub -l $__fish_git_prompt_branch_max_length $b)
+    if test $b_ori_length -gt $__fish_git_prompt_branch_max_length
+        set b "$b…"
```
It would probably be nice to have a `string shorten` subcommand that does all this. It could then incorporate some more logic, e.g. using `…` only if the current locale/encoding allows it.

And fish internally is _already_ doing that - see `ellipsis_char` in src/common.cpp.

Neonox31 commented:
Hi @faho,

Thanks for your time reviewing my pull request, taken your advices and completely reviewed it.
Modifications are the following :
- Added a "char suffix" var for changing defaulted ellipsis_char (your `string shorten` subcommand solution seems to be too complicated for me in this case because the ellispsis char is already used for untrackedfiles files with no detection of possible breaking locale/encoding)
- The "len" var has no more default value because not everybody wants to shorten the branch
- I check "len" var value for getting number format using `string match`
- Supressed useless "ori branch len" var
- Added some documentation in begining comments

I hope it will fit better than previous one.

Thanks.

faho commented on share/functions/__fish_git_prompt.fish:
```diff
@@ -416,10 +421,20 @@ function __fish_git_prompt --description "Prompt function for Git"
     if test -n "$b"
         set b "$branch_color$b$branch_done"
     end
+    if test -n "$b"
```
I am admittedly a bit sleepy, but why is this duplicated?

Neonox31 replied:
My mistake... sorry for that ! now fixed !

faho commented:
Merged, thanks!
